modules/gmc-agent/src/main/java/org/gridgain/action/Session.java

```diff
@@ -0,0 +1,213 @@
+
+    /** Session ID. */
+    private final UUID id;
+
+    /** Address. */
+    private InetSocketAddress addr;
+
+    /**
+     * Time when session is used last time.
+     * If this time was set at TIMEDOUT_FLAG, then it should never be changed.
+     */
+    private final AtomicLong lastTouchTime = new AtomicLong(U.currentTimeMillis());
+
+    /** Time when session was invalidated last time. */
+    private final AtomicLong lastInvalidateTime = new AtomicLong(U.currentTimeMillis());
+
+    /** Security context. */
+    private volatile SecurityContext secCtx;
+
+    /** Authorization context. */
+    private volatile AuthorizationContext authCtx;
+
+    /** Credentials that can be used for security context invalidation.*/
+    private volatile SecurityCredentials creds;
+
+    /**
+     * @param id Session ID.
+     */
+    private Session(UUID id) {
+        this.id = id;
+    }
+
+    /**
+     * Static constructor.
+     *
+     * @return New session instance with random session ID.
+     */
+    public static Session random() {
+        return new Session(UUID.randomUUID());
+    }
+
+    /**
+     * @return Session ID.
+     */
+    public UUID id() {
+        return id;
+    }
+
+    /**
+     * @return Security context.
+     */
+    public SecurityContext securityContext() {
+        return secCtx;
+    }
+
+    /**
+     * @param secCtx Security context.
+     */
+    public void securityContext(SecurityContext secCtx) {
+        this.secCtx = secCtx;
+    }
+
+    /**
+     * @return Authorization context.
+     */
+    public AuthorizationContext authorizationContext() {
+        return authCtx;
+    }
+
+    /**
+     * @param authCtx Authorization context.
+     */
+    public void authorizationContext(AuthorizationContext authCtx) {
+        this.authCtx = authCtx;
+    }
+
+    /**
+     * @return Security credentials.
+     */
+    public SecurityCredentials credentials() {
+        return creds;
+    }
+
+    /**
+     * @param creds Credentials.
+     */
+    public void credentials(SecurityCredentials creds) {
+        this.creds = creds;
+    }
+
+    /**
+     * @param time Time.
+     */
+    public void lastInvalidateTime(long time) {
+        lastInvalidateTime.set(time);
+    }
+
+    /**
+     * @return Remote client address.
+     */
+    public InetSocketAddress address() {
+        return addr;
+    }
+
+    /**
+     * @param addr Address.
+     */
+    public void address(InetSocketAddress addr) {
+        this.addr = addr;
+    }
+
+    /**
+     * Checks expiration of session and if expired then sets TIMEDOUT_FLAG.
+     *
+     * @param sesTimeout Session timeout.
+     * @return {@code True} if expired.
+     * @see #touch()
+     */
+    public boolean isTimedOut(long sesTimeout) {
+        long time0 = lastTouchTime.get();
+
+        if (time0 == TIMEDOUT_STATE)
+            return true;
+
+        return U.currentTimeMillis() - time0 > sesTimeout && lastTouchTime.compareAndSet(time0, TIMEDOUT_STATE);
```

Comment:
I would use `System.currentTimeMillis()` just in case.

Reply:
Fix.